src/zenml/services/service_registry.py

```diff
@@ -0,0 +1,179 @@
+
+    def register_service(self, service: "BaseService") -> None:
+        """Registers a new service instance
+
+        Args:
+            service: A BaseService instance.
+        """
+        service_type = service.SERVICE_TYPE
+        if service_type not in self.service_types:
+            raise TypeError("Service type <{service_type}> is not registered.")
+
+        if service.uuid not in self.services:
+            self.services[service.uuid] = service
+            logger.debug(f"Registered service {service}")
+        else:
+            existing_service = self.services[service.uuid]
+            raise Exception(
+                f"Found existing service {existing_service} for UUID: "
+                f"{service.uuid}. Skipping registration for service "
+                f"{service}."
+            )
+
+    def get_service(self, uuid: UUID) -> Optional["BaseService"]:
+        """Get the service instance registered for a UUID.
+
+        Args:
+            UUID: service instance identifier.
+
+        Returns:
+            `BaseService` instance that was registered for the UUID or
+            None, if no matching service instance was found.
+        """
+        return self.services.get(uuid)
+
+    def get_services(self) -> Dict[UUID, "BaseService"]:
+        """Get all service instances currently registered.
+
+        Returns:
+            Dictionary of `BaseService` instances indexed by their UUID with
+            all services that are currently registered.
+        """
+        return self.services.copy()
+
+    def service_is_registered(self, uuid: UUID) -> bool:
+        """Check if a service instance is registered."""
+        return uuid in self.services
+
+    def load_service_from_dict(
+        self, service_dict: Dict[str, Any]
+    ) -> "BaseService":
+        """Load a service instance from its dict representation.
+
+        Creates, registers and returns a service instantiated from the dict
+        representation of the service configuration and last known status
+        information.
+
+        If an existing service instance with the same UUID is already
+        present in the service registry, it is returned instead.
+
+        Args:
+            service_dict: dict representation of the service configuration and
+                last known status
+
+        Returns:
+            A new or existing ZenML service instance.
+        """
+        service_type = service_dict.get("service_type")
+        if not service_type:
+            raise ValueError(
+                "Service type not present in the service dictionary"
+            )
+        service_type = ServiceType.parse_obj(service_type)
+        service_class = self.get_service_type(service_type)
+        if not service_class:
+            raise TypeError(
+                f"Cannot load service with unregistered service "
+                f"type: {service_type}"
+            )
+        service = service_class.from_dict(service_dict)
```

Comment:
I'm a little confused how this works with pydantic, this `service_dict` still contains the service type which isn't part of the pydantic model right? Shouldn't we use `.pop(...)` in line 150?

Reply:
This was quite true, with the implementation at the time you reviewed these changes. But now, I'm adding `service_type` as an actual attribute part of the pydantic model, to replace all those `to_dict/to_json/from_dict/from_json` methods.